### PR TITLE
openjdk: update default version to 17.x

### PR DESCRIPTION
**Pull request by dialvarezs**
This PR updates the default version of OpenJDK from 11.x to the next LTS version, 17.x.

Measuring the full impact of this change is challenging. As an aproximation, I tested several packages that list Java as a dependency, focusing on those that do not specify a particular version and list the dependency as "build".

Here are the results. To test if a package "runs", I executed each program with `-h/--help` to verify if the output is correct and free of warnings. For some programs or libraries, I couldn't determine how to run or test them, so I marked those with a `?`. The only package in this list that builds with 11.x but not with 17.x is astral.

| package      | last_update | dependency_type | build | run |
|--------------|-------------|-----------------|-------|-----|
| ant          | 2024-03-12  | default         | yes   | yes |
| astral       | 2022-04-11  | build,run       | no    | -   |
| bbmap        | 2019-07/09  | default         | yes   | yes |
| fastqc       | 2023-06-05  | run             | -     | yes |
| hdf5+java    | 2024-01-10  | build,run       | yes   | ?   |
| lucene       | 2023-03-08  | run             | -     | yes |
| opencv       | 2023-09-26  | default         | yes   | ?   |
| openmpi+java | 2024-04-11  | default         | yes   | ?   |
| solr         | 2023-04-11  | run             | -     | yes |
| spark        | 2024-05-18  | build,run       | yes   | ?   |
| snpeff       | 2022-04-26  | build,run       | yes   | yes |
| tassel       | 2022-11-03  | build,run       | yes   | ?   |

Based on these results, the change appears to be safe. For any packages having issues, specifying the required version of OpenJDK should be neccesary.

I also include the propagation of the fix in #30826 for the newer MacOS versions.

EDIT:
From the pipeline results, it also seems that stc doesn't build, so it will also require to limit the java version in the package (or add a newer version, perhaps).

### CODE REVIEW

**Comment by wdconinc** [comment]
I won't be able to review for the next two weeks.

**Comment by alecbcs** [comment]
@dialvarezs can you modify `astral` so that it depends on `openjdk@11` ?

**Comment by dialvarezs** [comment]
> @dialvarezs can you modify `astral` so that it depends on `openjdk@11` ?

Done

**Comment by alecbcs** [comment]
@spackbot run pipeline

**Review by alecbcs — APPROVED** [review]
Looks good to me. Thanks @dialvarezs

File: var/spack/repos/builtin/packages/openjdk/package.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
             "7d3ab0e8eba95bd682cfda8041c6cb6fa21e09d0d9131316fd7c96c78969de31",
             "https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.3%2B9/OpenJDK21U-jdk_aarch64_linux_hotspot_21.0.3_9.tar.gz",
         ),
-        "Darwin-aarch64": (
+        "Darwin-arm64": (
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/openjdk/package.py:30]: I know I'm late with this, but can you document in a comment on the pr why this change was necessary? Did the os name get changed or was this wrong before?

Comment (wdconinc) [reply var/spack/repos/builtin/packages/openjdk/package.py:30]: Nevermind, I see the link to https://github.com/spack/spack/pull/30826 now...
